### PR TITLE
feat(security): rate limiting, auth enforcement, CORS, input validation, prompt guards

### PR DESCRIPTION
**Pull request by saikrishna-dev**
## Summary

Defense-in-depth security hardening (Feature 11). The backend was previously unprotected at the network layer — auth was frontend-only, zero rate limiting, and `/chat` was fully injectable.

**What changed:**

- **Rate limiting** (`slowapi`) — all endpoints protected; compute endpoints 5–20 req/min, readonly 60 req/min. Single `Limiter` instance in `dependencies.py`. 429 responses include `Retry-After` header and never leak internal detail.
- **Auth enforcement** — `require_user` FastAPI dependency added; applied to `POST /trade-setup` and `POST /jury/reanalyze` (HTTP 401 without a valid Bearer token). `SUPABASE_JWT_SECRET` controls signature verification; missing key logs a startup WARNING but allows dev-mode decode.
- **CORS** — `CORSMiddleware` restricted to `ALLOWED_ORIGINS` env var (defaults to both prod domains + localhost).
- **Input validation** — symbol regex `[A-Za-z0-9.\-:]{1,15}` enforced on `/predict`, `/dcf/{symbol}`, `/options/{symbol}`, and `/history`. Returns HTTP 422 on violation.
- **Prompt-injection guards** — `/chat` message capped at 500 chars; all context values sanitized (control chars stripped, 200-char cap); hardened system prompt; raw Groq error codes never forwarded to client.
- **Frontend auth forwarding** — all four Next.js proxy routes (`predict`, `chat`, `trade-setup`, `jury/reanalyze`) forward the `Authorization: Bearer` header. `AnalystJuryPanel`, `StockChatPanel`, and the analysis page pass `session.access_token`.
- **Observability** — 6 silent `except Exception: pass/continue` blocks upgraded to `logger.debug(...)`.

## Test plan

- [x] `python -m pytest backend/tests/ -v` — **99/99 pass** (10 new security tests + 89 existing)
- [x] `ruff check backend/` — clean
- [x] `pnpm run build` — clean TypeScript compile
- [x] `test_trade_setup_requires_auth` — 401 without Bearer token
- [x] `test_jury_reanalyze_requires_auth` — 401 without Bearer token
- [x] `test_predict_rejects_invalid_symbol` — 422 on `; DROP TABLE stocks--`
- [x] `test_predict_rejects_symbol_too_long` — 422 on 17-char symbol
- [x] `test_predict_accepts_valid_symbols` — AAPL, BRK.B, BTC-USD, AAPL:NASDAQ all pass
- [x] `test_dcf_rejects_invalid_symbol` — 422
- [x] `test_options_rejects_invalid_symbol` — 422
- [x] `test_chat_rejects_oversized_message` — 422 on 501-char message
- [x] `test_chat_accepts_500_char_message` — boundary condition passes validation
- [x] `test_rate_limit_returns_429_and_retry_after` — infrastructure verified

## Notes

`exempt_when` in slowapi is a zero-argument callable (no request context). The dual-limit tiered approach was replaced with a single `RATE_LIMIT_PREDICT_AUTH` limit using `_user_rate_key` (user UUID for authed callers, IP for anonymous) — this gives authed users their own isolated bucket without sharing an IP quota.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Global per-endpoint API rate limiting; stricter CORS allowlist; auth enforced for trade-setup and jury re-analysis; frontends forward auth headers.
  * Symbol input validation and chat sanitization with 500-char message cap; chat SSE surfaces clearer temporary-unavailable errors.

* **Behavior Change**
  * Analyst model updated to a smaller Llama 3.1 variant, affecting analyst verdicts/prompts.

* **Documentation**
  * Expanded Security section covering auth, rate limits, CORS, and input rules.

* **Tests**
  * Added security-focused tests validating auth, symbol validation, chat limits, and rate-limit responses.

### CODE REVIEW

File: .claude/FiForesight_Roadmap.md [hunk 1]
```diff
@@ -58,6 +58,7 @@
 - Router split — `main.py` → `routers/predict.py`, `simulation.py`, `trade.py`, `market.py` (#195)
 - App-tester QA fixes — jury consensus, DCF edge cases, SerpAPI, StockTwits, order book (#211)
 - Groq quota-exhaustion resilience — failure caching (5 min TTL), LangGraph fallback under New Relic (#208)
+- Security hardening — slowapi rate limits on all endpoints, auth enforcement (`require_user` on `/trade-setup` + `/jury/reanalyze`), CORS (`ALLOWED_ORIGINS`), symbol input validation (regex), `/chat` prompt-injection guards (500-char cap, control-char sanitization, hardened system prompt), JWT signature verification (`SUPABASE_JWT_SECRET`), observability logging on silent exceptions. 99 tests, ruff clean. (pending)
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply .claude/FiForesight_Roadmap.md]: Fixed: removed the stale (pending) suffix and replaced it with (#235).

File: backend/dependencies.py [hunk 2]
```diff
@@ -33,6 +41,51 @@
...
+        return ""
+    token = auth.removeprefix("Bearer ").strip()
+    if not token:
+        return ""
+    secret = Config.SUPABASE_JWT_SECRET
+    try:
+        if secret:
+            p = jwt.decode(token, secret, algorithms=["HS256"], options={"verify_aud": False})
+        else:
+            p = jwt.decode(token, options={"verify_signature": False})
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/dependencies.py:77]: Fixed: added ALLOW_INSECURE_JWT boolean to Config (default false). Both _extract_sub_fast and get_user_id now fail closed — when SUPABASE_JWT_SECRET is unset and ALLOW_INSECURE_JWT is not explicitly set to true, they return '' (anonymous) instead of decoding without verification. The startup warning is gated behind ALLOW_INSECURE_JWT=true.

File: backend/main.py [hunk 3]
```diff
@@ -30,7 +33,33 @@ async def lifespan(app: FastAPI):
...
+
+
+app.add_exception_handler(RateLimitExceeded, _rate_limit_handler)
+
+# --- CORS --------------------------------------------------------------------
+_origins = [o.strip() for o in Config.ALLOWED_ORIGINS.split(",") if o.strip()]
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=_origins,
+    allow_credentials=True,
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/main.py]: Fixed: added wildcard filter — _origins now excludes any '*' entries before passing to CORSMiddleware, with an error log if one is found.

File: backend/routers/trade.py [hunk 4]
```diff
@@ -214,23 +230,42 @@ async def trade_setup(req: TradeSetupRequest):
 # ---------------------------------------------------------------------------
 
 class ChatRequest(BaseModel):
-    message: str
+    message: str = Field(..., max_length=500)
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/routers/trade.py:233]: Fixed: req.message is now run through _CTRL_RE.sub(' ', req.message) before appending to messages, stripping control chars the same way context values are.

File: backend/routers/trade.py [hunk 5]
```diff
@@ -214,23 +230,42 @@ async def trade_setup(req: TradeSetupRequest):
...
+        current_price = _safe_ctx_value(ctx.get("currentPrice",  "N/A"), 30)
+        rsi_val      = _safe_ctx_value(ctx.get("rsi",            "N/A"), 20)
+        trend_val    = _safe_ctx_value(ctx.get("trend",          "N/A"), 20)
+        jury_summary = _safe_ctx_value(ctx.get("jury_summary",   "N/A"), 200)
+        sentiment    = _safe_ctx_value(ctx.get("sentiment_label","N/A"), 20)
+        headlines    = _safe_ctx_value(ctx.get("headlines",      "N/A"), 400)
+
+        # Validate symbol against the safe tag regex
+        if not _SYMBOL_RE.match(symbol.replace("N/A", "AAPL")):
+            symbol = "N/A"
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/routers/trade.py]: Fixed: replaced symbol.replace('N/A','AAPL') with an explicit sentinel check — if symbol != 'N/A' and not _SYMBOL_RE.match(symbol): symbol = 'N/A'.

File: backend/tests/test_routers.py [hunk 6]
```diff
@@ -4,18 +4,38 @@
...
+limiter = _deps.limiter
 
 _test_app = FastAPI()
+# Register the limiter (same instance as used by router decorators) so
+# slowapi can find it on request.app.state.
+_test_app.state.limiter = limiter
+_test_app.add_exception_handler(
+    RateLimitExceeded,
+    lambda req, exc: JSONResponse({"detail": "rate limited"}, status_code=429),
+)
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/tests/test_routers.py]: Fixed: replaced the lambda with a proper async handler that returns the same JSON body and Retry-After header as production.

File: backend/tests/test_security.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+def test_chat_accepts_500_char_message() -> None:
+    """Exactly 500 chars should be accepted (boundary condition)."""
+    # The streaming response will fail to reach Groq in test, but validation passes.
+    resp = _sec_client.post("/chat", json={
+        "message": "x" * 500,
+        "context": {"symbol": "AAPL"},
+        "history": [],
+    })
+    # Not 422 means validation passed (may fail downstream without a real Groq key)
+    assert resp.status_code != 422, "500-char message was incorrectly rejected"
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/tests/test_security.py:146]: Fixed: wrapped the _sec_client.post call in patch('httpx.AsyncClient.stream') so the test never makes a real network call; only Pydantic validation is exercised.

File: CLAUDE.md [hunk 8]
```diff
@@ -93,7 +93,7 @@ cd frontend && pnpm run dev     # Next.js on :3000
 python -m uvicorn main:app --app-dir backend --host 0.0.0.0 --port 8000 --reload
 
 # Tests
-python -m pytest backend/tests/ -v   # 22 tests, ~1s
+python -m pytest backend/tests/ -v   # 99 tests, ~5s
 ruff check backend/                   # linter
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply CLAUDE.md:97]: Fixed: updated both occurrences of '22 tests' to '99 tests' in CLAUDE.md (project-structure comment and shipped-items table).

File: backend/services.py [hunk 9]
```diff
@@ -1752,34 +1783,62 @@ async def call_groq_with_tools(
...
+                status = exc.response.status_code
+                # ── 429 rate-limit: wait Retry-After (capped at 8 s) then retry ──
+                if status == 429:
+                    retry_after = 6.0
+                    try:
+                        import re as _re
+                        m = _re.search(r"try again in (\d+(?:\.\d+)?)s", exc.response.text)
+                        if m:
+                            retry_after = min(float(m.group(1)) + 0.5, 8.0)
+                    except Exception:
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/services.py]: Fixed: replaced bare except Exception: pass with except Exception as parse_exc: logger.debug(...) so failures during Retry-After parsing are traceable without raising.

File: backend/services.py [hunk 10]
```diff
@@ -2026,8 +2087,15 @@ async def get_analyst_verdict(
             )
 
         except Exception as e:
+            # Log the HTTP response body for easier diagnosis of 4xx/429 errors
+            _body = ""
+            if isinstance(e, httpx.HTTPStatusError):
+                try:
+                    _body = f" | response_body={e.response.text[:300]}"
+                except Exception:
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/services.py]: response_body=<unavailable>'.

File: backend/services.py [hunk 11]
```diff
@@ -1752,34 +1782,62 @@ async def call_groq_with_tools(
...
+                status = exc.response.status_code
+                # ── 429 rate-limit: wait Retry-After (capped at 20 s) then retry ──
+                if status == 429:
+                    retry_after = 6.0
+                    try:
+                        import re as _re
+                        m = _re.search(r"try again in (\d+(?:\.\d+)?)s", exc.response.text)
+                        if m:
+                            retry_after = min(float(m.group(1)) + 0.5, 20.0)
+                    except Exception:
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/services.py]: Duplicate of the earlier comment — same fix applied: bare except replaced with logged fallback.

File: backend/services.py [hunk 12]
```diff
@@ -2026,8 +2086,15 @@ async def get_analyst_verdict(
             )
 
         except Exception as e:
+            # Log the HTTP response body for easier diagnosis of 4xx/429 errors
+            _body = ""
+            if isinstance(e, httpx.HTTPStatusError):
+                try:
+                    _body = f" | response_body={e.response.text[:300]}"
+                except Exception:
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/services.py]: Duplicate of the earlier comment — same fix applied: bare except replaced with logged fallback.

File: backend/services.py [hunk 13]
```diff
@@ -1752,34 +1799,62 @@ async def call_groq_with_tools(
...
+                status = exc.response.status_code
+                # ── 429 rate-limit: wait Retry-After (capped at 20 s) then retry ──
+                if status == 429:
+                    retry_after = 6.0
+                    try:
+                        import re as _re
+                        m = _re.search(r"try again in (\d+(?:\.\d+)?)s", exc.response.text)
+                        if m:
+                            retry_after = min(float(m.group(1)) + 0.5, 20.0)
+                    except Exception:
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/services.py]: Already fixed in commit cdb9b2d: replaced bare `except Exception: pass` with `except Exception as parse_exc: logger.debug(...)` — logs retry-after parse failures at DEBUG level while preserving the default-backoff fallback behaviour.

File: backend/services.py [hunk 14]
```diff
@@ -2026,13 +2103,27 @@ async def get_analyst_verdict(
             )
 
         except Exception as e:
+            # Log the HTTP response body for easier diagnosis of 4xx/429 errors
+            _body = ""
+            if isinstance(e, httpx.HTTPStatusError):
+                try:
+                    _body = f" | response_body={e.response.text[:300]}"
+                except Exception:
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/services.py]: Already fixed in commit cdb9b2d: replaced bare `except Exception: pass` with `except Exception as body_err: logger.debug(...); _body = '| response_body=<unavailable>'` — non-fatal, but now self-documenting and debuggable.
